include/deal.II/base/task_result.h

```diff
@@ -504,10 +517,44 @@ namespace Threads
 
 
 
+  template <typename T>
+  inline bool
+  TaskResult<T>::empty() const
+  {
+    // If we have waited before, then return immediately:
+    if (result_is_available)
+      return false;
+    else
+      // If we have not waited, wait now. We need to use the double-checking
+      // pattern to ensure that if two threads get to this place at the same
+      // time, one returns right away while the other does the work. Note
+      // that this happens under the lock, so only one thread gets to be in
+      // this code block at the same time:
+      {
+        std::lock_guard<std::mutex> lock(mutex);
+        if (result_is_available)
+          return false;
+        else
+          // If there is a task, then this object is not empty. Otherwise, we
+          // have no result and no task, so the object is empty:
+          if (task.has_value())
+            return false;
+          else
+            return true;
+      }
```

Comment:
I guess this is a bit of a refactoring artifact: I am not sure why the logic is not a simple:
```
if (result_is_available)
  return false;
else if (task.has_value())
    return false;
 else
    return true;
```

The double checking that you are doing here doesn't prevent concurrent threads from entering the last `else` branch with the `task.has_value()` because that has nothing to do with `result_is_available` and I don't think we need to guard the `task.has_value()`, it should be thread safe. Or I am missing something :sweat_smile:

Reply:
Well, that's not quite right. The function that sets `result_is_available` (namely, `join()`) clears `task`. So they are related. 

The situation I want to cover is where you set a task and at a later time, one thread calls `value()` (which implicitly calls `join()`) and another thread calls `empty()`. `empty()` should return `false` here. There are two cases:
* `empty()` before `join()`: `result_is_available` is `false`, but `task.has_value()` is `true`.
* `empty()` after `join()`: `result_is_available` is `true`, but `task.has_value()` is `false`.
The issue is that while `join()` is running, we do not have any guarantees about the relative states of these two variables, and if I don't use the lock, `empty()` will return true or false depending on the order of things done in `join()`.

For reference, my mental model for what needs to be thread-safe is that functions that are `const` can be called concurrently and should work. Both the current function `empty()` and `join()`/`value()` are `const`, and so I would like to be able to call them concurrently. I think the logic above guarantees that, but only with the critical section guarded by the mutex.

Reply:
@bangerth I understand now what you try to do - and it reminds me that multithreading and concurrency is hard...

I still think that taking the lock here is expensive and unnecessary. If `join()`ing the result takes 5 minutes then I really don't want `empty()` to block for 5 minutes.

What about simply not clearing the `std::optional` `task` in `join()`?

That way `task.has_value()` properly functions as a boolean and you could simply implement
```
bool empty()
{
  return !task.has_value();
}
```

Reply:
And this is then thread safe in your mental model because everything that can change `task.has_value()` are non-`const` functions.

Reply:
Hard: ha, yes. Plenty of mistakes to be made :-)

Joining taking 5 minutes: You're right that if you call `join()` first, then you wait in `empty()` until the task has finished. That's not great. What if indeed I start the function as in the version I just pushed, where we really go under the lock if we apparently are racing `join()`'s resetting of member variables? In particular, just because `join()` is currently waiting does not cause us to wait here also.

Reply:
@bangerth The current version looks good so let's get that merged. I am not going to start mumbling "lock contention" in fear of getting chased with a pitchfork. I still think we should simply not clear the `std::optional` `task` object and simply use `task.has_value()` :smile:

Reply:
I did consider this, but it would mean that we keep the `Task` data structure around forever. In the case of what we were discussing for the `FE_*` classes' restriction and prolongation matrices, that would typically mean for the entire lifetime of the program.